torch/CMakeLists.txt

```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
     ${TORCH_SRC_DIR}/csrc/api/src/optim/lbfgs.cpp
     ${TORCH_SRC_DIR}/csrc/api/src/optim/rmsprop.cpp
     ${TORCH_SRC_DIR}/csrc/api/src/optim/sgd.cpp
-  )
+    )
+
 endif()
 
 add_library(torch SHARED ${TORCH_SRCS})
 
-# https://gcc.gnu.org/onlinedocs/gcc-4.0.3/gcc/Warning-Options.html
-target_compile_options(torch
-  PRIVATE
-  -Wall
-  -Wextra
-  -pedantic
-  -Wcast-align
-  -Wcast-qual
-  -Wctor-dtor-privacy
-  -Wdisabled-optimization
-  -Winit-self
-  -Wmissing-include-dirs
-  -Woverloaded-virtual
-  -Wsign-promo
-  -Wstrict-overflow=5
-  -Wundef
-  -fdiagnostics-show-option
-  -Wno-unused-parameter
-  -Wno-missing-braces # This warning is buggy
-  -Wno-unknown-pragmas)
-
-if ($ENV{WERROR})
+# until they can be unified, keep these lists synced with setup.py
+if(MSVC)
+  target_compile_options(torch PRIVATE
+    /MD
+    /Z7
+    /EHa
+    /DNOMINMAX
+    /wd4267
+    /wd4251
+    /wd4522
+    /wd4522
+    /wd4838
+    /wd4305
+    /wd4244
+    /wd4190
+    /wd4101
+    /wd4996
+    /wd4275
+    /bigobj
+    )
+else()
+  target_compile_options(torch PRIVATE
+    -std=c++11
+    -Wall
+    -Wextra
+    -Wno-unused-parameter
+    -Wno-missing-field-initializers
+    -Wno-write-strings
+#    -Wno-zero-length-array
+    -Wno-unknown-pragmas
+    # This is required for Python 2 declarations that are deprecated
+    # in 3.
+    -Wno-deprecated-declarations
+    # Python 2.6 requires -fno-strict-aliasing see
+    # http://legacy.python.org/dev/peps/pep-3123/ We also depend on it
+    # in our code (even Python 3).
+    -fno-strict-aliasing
+    # Clang has an unfixed bug leading to spurious missing braces
+    # warnings, see https://bugs.llvm.org/show_bug.cgi?id=21629
+    -Wno-missing-braces
+    )
+endif()
+
+if (MSVC)
+elseif ($ENV{WERROR})
   target_compile_options(torch PRIVATE -Werror)
 endif()
 
+if (MSVC)
```

Comment:
I don't understand: why is this MSVC only?

Reply:
windows doesn't have implicit transitive shared libraries. The setup.py build does the same, with this comment (which i can copy over)

    # Windows needs direct access to ONNX libraries as well
    # as through Caffe2 library

Reply:
It's better practice to not rely on implicit transitive shared libraries, even on Linux.

Reply:
ok, I'll send a test run to see it I can remove the `if (MSVC)`. If it doesn't work though, I claim it's out of scope of this diff to invest more time changing it, since the behavior here is unchanged (and was intentional according to that comment).

Reply:
does not work out of the box https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-py3-clang5-asan-test1/3865/console